### PR TITLE
JAMES-33775 RSpamDListener

### PR DESCRIPTION
**Pull request by vttranlina**
jira https://issues.apache.org/jira/browse/JAMES-3775

### CODE REVIEW

**Review by Arsnael — APPROVED** [review]

File: mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    private Mono<Void> reportHamWhenAdded(MailboxEvents.Added addedEvent, MailboxSession session) {
+        return mapperFactory.getMailboxMapper(session)
+            .findMailboxById(addedEvent.getMailboxId())
+            .map(mailbox -> Pair.of(mailbox, mapperFactory.getMessageMapper(session)))
+            .flatMapMany(pair -> Flux.fromIterable(MessageRange.toRanges(addedEvent.getUids()))
+                .flatMap(range -> pair.getRight().findInMailboxReactive(pair.getLeft(), range, MessageMapper.FetchType.FULL, LIMIT)))
+            .map(Throwing.function(MailboxMessage::getFullContent))
+            .map(this::getHeaderAndContentInputStream)
+            .flatMap(headerAndContent -> rSpamDHttpClient.reportAsHam(headerAndContent.getLeft(), headerAndContent.getRight()), ReactorUtils.DEFAULT_CONCURRENCY)
```
Comment (quantranhong1999) [inline mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java:123]: IMO:
```suggestion
            .flatMap(mailboxMessage -> rSpamDHttpClient.reportAsHam(mailboxMessage.getHeaderContent(), mailboxMessage.getBodyContent()), ReactorUtils.DEFAULT_CONCURRENCY)
```

File: mailbox/plugin/rspamd/pom.xml [hunk 2]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline mailbox/plugin/rspamd/pom.xml:1]: No IMO keep this in `third-part/rspamd` ...

File: mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java:123]: IMO all of this code would be way easier if we can `rSpamDHttpClient.reportAsHam(fullContent)` and `rSpamDHttpClient.reportAsSpam(fullContent)`.

Now we are parsing the mime message to split header and content to merge it again later down the road, which is complex and inefficient.

Can we do this refactoring please?

File: mailbox/pom.xml [hunk 3]
```diff
@@ -57,6 +57,7 @@
         <module>plugin/quota-search-elasticsearch-v7</module>
         <module>plugin/quota-search-scanning</module>
 
+        <module>plugin/rspamd</module>
```
Comment (chibenwa) [inline mailbox/pom.xml:60]: Nope let's not create such a plugin...

**Review by chibenwa — COMMENTED** [review]
If we are interestead about consistency we can propose on the mailing list to move all spamassassin stuff in a new `thrid-party/spamassassin` module.

WDYT?

Can you start such a thread?

File: mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java (hunk 1 above)
Comment (vttranlina) [reply mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java:123]: Good job, it's pretty

Comment (chibenwa) [reply mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java:123]: CF https://github.com/apache/james-project/pull/1100

**Comment by vttranlina** [comment]
> If we are interestead about consistency we can propose on the mailing list to move all spamassassin stuff in a new `thrid-party/spamassassin` module.

Yes, 
https://www.mail-archive.com/server-dev@james.apache.org/msg72205.html

**Comment by vttranlina** [comment]
Rebase master & squash fixup

**Review by Arsnael — APPROVED** [review]

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDListener.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        return handleMessageMoved(mailboxMessagePublisher(messageMoveEvent), messageMoveEvent);
+    }
+
+    private Mono<Void> reportHamWhenAdded(MailboxEvents.Added addedEvent, MailboxSession session) {
+        return mapperFactory.getMailboxMapper(session)
+            .findMailboxById(addedEvent.getMailboxId())
+            .map(mailbox -> Pair.of(mailbox, mapperFactory.getMessageMapper(session)))
+            .flatMapMany(pair -> Flux.fromIterable(MessageRange.toRanges(addedEvent.getUids()))
+                .flatMap(range -> pair.getRight().findInMailboxReactive(pair.getLeft(), range, MessageMapper.FetchType.FULL, LIMIT)))
+            .map(mailboxMessage -> Throwing.supplier(mailboxMessage::getFullContent).get())
```
Comment (quantranhong1999) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDListener.java:116]:
```suggestion
            .map(Throwing.function(MailboxMessage::getFullContent))
```

File: third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDListener.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            .flatMap(rSpamDHttpClient::reportAsHam, ReactorUtils.DEFAULT_CONCURRENCY)
+            .then();
+    }
+
+    private Flux<InputStream> mailboxMessagePublisher(MessageMoveEvent messageMoveEvent) {
+        return Mono.fromCallable(() -> mapperFactory.getMessageIdMapper(mailboxManager.createSystemSession(Username.of(getClass().getCanonicalName()))))
+            .flatMapMany(messageIdMapper -> messageIdMapper.findReactive(messageMoveEvent.getMessageIds(), MessageMapper.FetchType.FULL))
+            .map(mailboxMessage -> Throwing.supplier(mailboxMessage::getFullContent).get());
```
Comment (quantranhong1999) [inline third-party/rspamd/src/main/java/org/apache/james/rspamd/RSpamDListener.java:124]:
```suggestion
            .map(Throwing.function(MailboxMessage::getFullContent));
```

**Pull request merged**